ledger-core/inc/core/operation/OperationDatabaseHelper.inl

```diff
@@ -0,0 +1,191 @@
+#include <core/database/SociDate.hpp>
+#include <core/database/SociOption.hpp>
+#include <core/collections/Strings.hpp>
+#include <core/operation/Operation.hpp>
+#include <core/wallet/BlockDatabaseHelper.hpp>
+#include <core/wallet/TrustIndicator.hpp>
+
+namespace ledger {
+    namespace core {
+
+        namespace impl {
+
+            // Declares a type trait to check if templated type has a static method
+            // called `updateCurrencyOperation`.
+            template <typename T, typename = void>
+            struct has_update;
+
+            // Uses SFINAE to determine if templated type has the 
+            // `updateCurrencyOperation` method.
+            // If the substitution succeed, `has_update` is evaluated as a
+            // `true_type`
+            template <typename T>
+            struct has_update<T, decltype(T::updateCurrencyOperation)> : std::true_type
+            {};
+
+            // Else the `has_update` is evaluated as a `false_type`
+            template <typename T>
+            struct has_update<T, void> : std::false_type
+            {};
+
+            // Syntaxic sugar here to avoid to write the `has_update<T>::value`
+            template <typename T>
+            constexpr auto has_update_v = has_update<T>::value;
+        }
+
+        template <typename Derived>
+        bool OperationDatabaseHelper<Derived>::putOperation(soci::session& sql, const Operation& operation)
+        {
+            static_assert(
+                impl::has_update_v<Derived>,
+                "OperationDatabaseHelper<T> requires 'T' to define 'updateCurrencyOperation' method");
+
+            using namespace soci;
+
+            auto count = 0;
+            std::string serializedTrust;
+            // TODO: I'm honestly not sure at all about this line below - since
+            // api::TrustIndicator has none `serialize` method, we obviously need the
+            // underlying implementation. Change `api::TrustIndicator` to
+            // `TrustIndicator` in the `Operation` class is for sure a better solution 
+            serialization::saveBase64<TrustIndicator>(*dynamic_cast<TrustIndicator*>(operation.trust.get()), serializedTrust);
+            if (operation.block.nonEmpty()) {
+                BlockDatabaseHelper::putBlock(sql, operation.block.getValue());
+            }
+            auto blockUid = operation.block.map<std::string>([] (const api::Block& block) {
+                return block.uid;
+            });
+            sql << "SELECT COUNT(*) FROM operations WHERE uid = :uid", use(operation.uid), into(count);
+            auto newOperation = count == 0;
+            if (!newOperation) {
+                sql << "UPDATE operations SET block_uid = :block_uid, trust = :trust WHERE uid = :uid"
+                        , use(blockUid)
+                        , use(serializedTrust)
+                        , use(operation.uid);
+                Derived::updateCurrencyOperation(sql, operation, newOperation);
```

Comment:
Why should `OperationDatabaseHelper` depends on Derived?
Can the Derived do something like 
```
bool Derived::putOperation(soci::session& sql, const Operation& operation) {
  updateCurrencyOperation(sql, operation, OperationDatabaseHelper::putOperation(sql, operation));
}
```
?
This way there should be no dependencies for Derived in OperationDatabaseHelper, so it can be non-template

Reply:
You make a point here ! In the current state, there are no problem with your solution because of the implementation of `putOperation` but there are some pitfalls I want to avoid.  

First I cannot guarantee than the `putOperation` definition will never change - we have to change all the derived classes if such a thing happens
Second I choose CRTP because all methods are statics and there are no way to tell to the derived class to implement/override this `putOperation` - this is error-proned and unclear. By taking a derived class as templated type, we mandatory the definition of a static method named `updateCurrencyOperation`.

Reply:
> First I cannot guarantee than the putOperation definition will never change - we have to change all the derived classes if such a thing happens
If it changes, for some currencies it may not works correctly anymore.

> Second I choose CRTP because all methods are statics and there are no way to tell to the derived class to implement/override this putOperation - this is error-proned and unclear. By taking a derived class as templated type, we mandatory the definition of a static method named updateCurrencyOperation
CRTP is better then inheritance. But simple function composition is better then CRTP. Less dependencies, the better.